source/s3_client.c

```diff
@@ -1168,20 +1168,34 @@ static struct aws_s3_meta_request *s_s3_client_meta_request_factory_default(
     /* Call the appropriate meta-request new function. */
     switch (options->type) {
         case AWS_S3_META_REQUEST_TYPE_GET_OBJECT: {
-            /* If the initial request already has partNumber, the request is not
-             * splittable(?). Treat it as a Default request.
-             * TODO: Still need tests to verify that the request of a part is
-             * splittable or not */
-            if (aws_http_headers_has(initial_message_headers, aws_byte_cursor_from_c_str("partNumber"))) {
-                return aws_s3_meta_request_default_new(
-                    client->allocator,
-                    client,
-                    AWS_S3_REQUEST_TYPE_GET_OBJECT,
-                    content_length,
-                    false /*should_compute_content_md5*/,
-                    options);
+            struct aws_byte_cursor path_and_query;
+
+            if (aws_http_message_get_request_path(options->message, &path_and_query) == AWS_OP_SUCCESS) {
+                /* If the initial request already has partNumber, the request is not
+                 * splittable(?). Treat it as a Default request.
+                 * TODO: Still need tests to verify that the request of a part is
+                 * splittable or not */
+                struct aws_byte_cursor sub_string;
+                AWS_ZERO_STRUCT(sub_string);
+                /* The first split on '?' for path and query is path, the second is query */
+                if (aws_byte_cursor_next_split(&path_and_query, '?', &sub_string) == true) {
+                    aws_byte_cursor_next_split(&path_and_query, '?', &sub_string);
+                    struct aws_uri_param param;
+                    AWS_ZERO_STRUCT(param);
+                    struct aws_byte_cursor part_number_query_str = aws_byte_cursor_from_c_str("partNumber");
+                    while (aws_query_string_next_param(sub_string, &param)) {
+                        if (aws_byte_cursor_eq(&param.key, &part_number_query_str)) {
+                            return aws_s3_meta_request_default_new(
+                                client->allocator,
+                                client,
+                                AWS_S3_REQUEST_TYPE_GET_OBJECT,
+                                content_length,
+                                false /*should_compute_content_md5*/,
+                                options);
+                        }
+                    }
```

Comment:
Debatable: I think it would be better to invest in a better helper like `aws_http_headers_has` or `aws_http_headers_get`. This logic will need to be repeated wherever we need to extract a single parameter from query params.

Or, we could still use `find` and try to locate `&partNumber=` or `?partNumber=`.

Reply:
I have no strong feelings about it.

My take is we are used as a client level http lib, and this is the first time we ever really want to parse the query params, maybe let's hold it back until the next time we want to reuse it?